src/catalog/database_catalog.cpp

```diff
@@ -371,13 +371,46 @@ bool DatabaseCatalog::DeleteNamespace(transaction::TransactionContext *const txn
     return false;
   }
 
-  // Step 4: Delete from oid index
+  // Step 4: Cascading deletes
+  // Get the objects in this namespace
+  auto ns_objects = GetNamespaceObjectOids(txn, ns_oid);
+  for (const auto object : ns_objects) {
+    // Delete all of the tables. This should get most of the indexes
+    if (object.second == postgres::ClassKind::REGULAR_TABLE) {
+      result = DeleteTable(txn, static_cast<table_oid_t>(object.first));
+      if (!result) {
+        // Someone else has a write-lock. Free the buffer and return false to indicate failure
+        delete[] buffer;
+        return false;
+      }
+    }
+  }
+
+  // Get the objects in the namespace again, just in case there were any indexes that don't belong to a table in this
+  // namespace. We could do all of this cascading cleanup with a more complex single index scan, but we're taking
+  // advantage of existing PRIs and indexes and expecting that deleting a namespace isn't that common of an operation,
+  // so we can be slightly less efficient than optimal.
+  ns_objects = GetNamespaceObjectOids(txn, ns_oid);
+  for (const auto object : ns_objects) {
+    // Delete all of the straggler indexes that may have been built on tables in other namespaces. We shouldn't get any
+    // double-deletions because indexes on tables will already be invisible to us (logically deleted already).
+    if (object.second == postgres::ClassKind::INDEX) {
+      result = DeleteIndex(txn, static_cast<index_oid_t>(object.first));
+      if (!result) {
+        // Someone else has a write-lock. Free the buffer and return false to indicate failure
+        delete[] buffer;
+        return false;
+      }
+    }
+  }
+
```

Comment:
It is probably worth adding a third fetch inside an assert here that ensures the results are empty.  This will catch new object types (like views) as we add them in the future.